### ISSUE TITLE
General questions regarding the paper 

### ISSUE THREAD

**Issue by kai981** [facebookresearch/DeltaCNN#7]
Hi, I am back with more questions :)

1. In Section 2.2, it was stated that “we use a spatial (per pixel) sparsity, since structured sparsity better suits SIMD architectures than per value sparsity”, what does “per value sparsity” refer to here? 

2. In Section 3.2, it was stated that a benefit of using tiled convolutions is that “input features and filter parameters can be kept local and reused multiple times”, could you elaborate more on this? 

3. In Section 3.3, it was stated that “we also need to limit the increase in accuracy when tuning for thresholds to avoid overfitting n the small subset of the training set”, why would there be increase in accuracy when increasing threshold would result in increase of loss? 

4. In the supplementary material, Section 1.2, it was stated that “This mode loads only pixels of the filter weights that are required and iterates over an array of active pixels contrary to iterating over all pixels and checking the update flag”, could you elaborate more on this? For very sparse mode, if checking of update flag is not needed, how does it know which pixel is updated? 

5. In the same section, regarding depth-wise convolutions, it was stated that “input values are used less often and only by a single thread, removing the necessity of keeping data local to a CTA”, what does it mean by keeping data local to a CTA? 
Depth-wise convolutions use per-pixel sparse operation, could you elaborate more on this? So it decides pixel by pixel if update is needed, if it is needed, then convolution is performed on this one pixel? 

6. Regarding results in Table 2, for RTX 3090 b=1, are the results obtained using a single GPU?
I have also observed that, for MOT16-d1, GTX 1050 b=1, the difference between FPS of cuDNN and ‘ours dense’ is 10.5 vs 11.6 (a rather small difference), but a large difference when threshold is tuned, which is ‘ours sparse’, FPS of 27.4. However, for RTX 3090 b=1, the different between cuDNN and ‘ours dense’ is rather large, 23.3 vs 45.0, but further tuning threshold only increases FPS to 45.2. Is there any reasons for such observations? To clarify, for ‘ours dense’ and ‘ours sparse’, both are performed in sparse operations, but ‘ours dense’ has negative thresholds so no values are truncated, while ‘ours sparse’ has tuned thresholds, is that right? 

Thank you so much!

**Comment by dabeschte** [comment]
Hey!
Sure, but this might become a lengthy response ;)

1: RRM (recurrent residual module) uses a per value sparsity in the sense of: they assume to achieve speedups by even only skipping the multiplication of a single value. In reality, the overhead of checking a value against zero is often higher than the gains of avoiding a multiplication. With our per pixel sparsity, we check only once per pixel if all channels will be processed, or all of the can be skipped at once. 

2: when you perform a 3x3 convolution for one pixel, you need to load 9 input pixels and 9 filter values. when processing the neighbor pixel, you again need to load 9 pixels (6 of which are the same as in pixel before) and 9 filter valued (all of them are the same as before). If we process the image in tiles of 2x1 pixels, we only need to load 9+9+3+0 pixels = 21. processing every pixel individually requires 9+9+9+9 loading operations. Because of that, it makes sense to process as many pixels as possible in tiles to maximize data locality and thereby decrease redundant loading operations. however, the tile size is in the end limited by the number of registers available per thread on a GPU.

3: in some cases it can actually happen that a slight increase of one of the thresholds leads to a decrease in loss. Of course, if you then keep increasing it, the loss will increasing again at some point. In order to keep as close as possible to "vanilla" inference, we also need to prevent significant accuracy increase with individual thresholds.

4: the sparse mode does still check the update mask. but I understand that this sentence is indeed confusing. what I meant here is that during the multiplication, we don't iterate over all 8x8 inputs pixels to process the 6x6 output pixels. instead, we only iterate over the (up to 4) of the 8x8 pixels that we know are active. this avoid the extensive checks of the update mask. while this sounds like a really minor optimization, it actually saves a lot of instructions: since we store the update mask as a binary mask to save registers, we need to perform a lot of bit operations for every check if a pixel is updated. When you have to process 256 channels, iterating over 64 inputs and checking the update mask becomes expensive. So, what we do here instead is to store the (up to) 4 active pixels and their locations and only iterate over them.

5: we always process one output channel per thread. a CTA consists of many threads, with the number depending on the number of channels that we process. let's say, we process a normal convolution (not depth-wise) with 128 input channels and 128 output channels. using a CTA with 128 threads, every threads produces 1 output channel, but has to load all 128 input pixels to calculate the result. in depth-wise convolutions, the result of one output channel only depends on a single input channel. for normal convolutions, it makes sense to save inputs in shared memory as they are reused by all threads. for depth-wise convolutions, it does not make sense to use shared memory - we instead load the values during the multiplication loop. That also means that we need to check the update mask during the multiplication loop and therefore can also just skip loading and multiplying an input pixel entirely. So yes, your explanation is correct.

6: yes, the difference here is that the 1050 is fully utilized with dense inference with a batch size of 1. adding sparsity does help here. the 3090 is only partially utilized with a batch size of 1. Adding sparsity does not make a difference. The reason why we are so much faster with b=1 in dense mode is quite interesting. But I think the reason is that PyTorch/cuDNN add a lot of overhead on the cpu while we are very low overhead and well optimized for the use case of low utilization. but that is just a guess. Another reason could be that back then, PyTorch/cuDNN automatically performed 16-bit tensor core convolutions when possible, even when 32-bit was used for model and input. This might add some overhead, but gave them a big advantage with higher batch sizes. In later PyTorch versions, cuDNN is now correctly using 32 bit convolutions. this levels the playing field and increases the speedup of our approach relative to cuDNN. Also, yes: ours dense and ours sparse are implemented as you described.


I hope this explanation helps. It did get quite lengthy indeed ;)

**Comment by kai981** [comment]
Hi, really appreciate your efforts to give such elaborate answers! :) 
I have a follow-up question and some new questions regarding implementation. 

**Follow-up**
According to your explanation for (1), I think there might be a minor typo in the paper if I am not wrong, as the sentence after the one I enquired about is **"Like RRM, we decide per input pixel if an update is required."**, I guess it should be **"Unlike RRM"** instead?

**Implementation questions**
1. I am trying to perform inference on the few MOT16 sequences with fixed camera. If I process these sequences individually, it is producing similar AP for original model and dc model. However, if I try to process all these sequences in a loop one by one, by adding `dc_model.reset_layers()` at the end of the loop, it does not seem to be producing the right results, the AP for dc model drops drastically. I am suspecting that the model is reset wrongly or I am missing some steps, may I know what is the correct way to reset the model?

2. Another issue is that, the current model I have has a Head layer that returns a list of five feature maps of different scales as output. If I create a unique instance for the last point-wise convolution layer of each scale and input the argument dense_out=True, it is able to produce the correct results. However, I would have to modify the state_dict by duplicating the convolution weights for each of the unique instance. So I tried another way, which is to use the same instance for the last point-wise convolution so the output for the Head layer is a list of five feature maps in sparse mode. Then, I created unique instances of Densify() layer: `self.dclass_rep = nn.ModuleList([DCDensify() for _ in range(5)])` (similarly for bbox output). In the forward function I would have: 
```
        x_class = self.class_net(x)
        x_box = self.box_net(x)
        for i, (densify_class, densify_box) in enumerate(zip(self.dclass_rep, self.dbox_rep)): 
            x_class[i] = densify_class(x_class[i])
            x_box[i] = densify_box(x_box[i])
        return x_class, x_box 
```
where class_net and box_net are the Head layers, so x_class and x_box are lists of feature maps in sparse mode. 
This second way is not able to produce the correct results, could you advice on what might be the issue?

Thank you!

**Comment by dabeschte** [comment]
You are welcome :)

regarding "Like RRM, we decide per input pixel if an update is required.": true, that is a confusing sentence. What I meant is that we decide for each INPUT whether it requires processing or not. Skip-Conv decides that per OUTPUT. But the "term" pixel here is of course wrong, because RRM uses per-input-value sparsity. we use per-input-pixel sparsity and RRM uses per-output-pixel sparsity.

regarding implementation:
1: your approach is correct. it seems like some part is not correctly reset. Are the sparsify and densify layer at the beginning and the end part of the model and therefore also correctly reset? Are all classes that use DeltaConv layers derived from DCModule (or nn.Module should also work). Are all layers initialized in the __init__ function or are some initialized at a later point? because then it could happen that pytorch's module.modules() function misses them. If you struggle to find the error, you can use the internal logger to log a lot of information in the convolutional layers. 
I unfortunately did not document it, because this interface is not really polished. But something like this should work:
```
from torch.utils.tensorboard import SummaryWriter
summary_writer = SummaryWriter(<log_path>)
LoggingController.img_epochs = 1
log_contr = LoggingController(model, summary_writer)
DCConv2d.use_logging = True
log_contr.reset_loggers(enable=True)
for frame_idx, frame in enumerate(frames):
  out = model(frame)
  log_contr.write_logs(frame_idx, reset=True, disable=False)
```

2: I had to dig through my code to find a potential problem here. I think the problem is that the convolution layer keeps track of the frame idx and it only applies the bias on the first frame (see [here](https://github.com/facebookresearch/DeltaCNN/blob/07f49061483f888752df842b461d8a85ed7c062a/src/deltacnn/sparse_layers.py#L617)). In your case, you'd need to apply the bias the first 5 times. When you use 5 instances of the layer, all of them are only called once in the first frame. Can you please check if this layer uses a bias? If so, you could e.g. reset the frame_idx to -1 before all of the 5 calls in the first frame. But be careful, because you have to apply biases again once 'reset_layers' is called

**Comment by kai981** [comment]
Hi, some updates here. 

1. I have tried to run the following check
```
            dc_model.reset_layers() 
            for mod in dc_model.modules():
                try:
                    print(mod.idx)
                    print(mod)
                except AttributeError:
                    try:
                        print(mod.conv_idx)
                        print(mod)
                    except AttributeError:
                        continue
                    continue 
```
I realized that the densify layers are all not reset properly, and these layers are all called in DCConv2d using dense_out=True instead of having unique instances by themselves in the dc_model (not sure if this might be an issue). I don't think module.modules() function missed them, since I am using that to check the idx and it's able to print out those DCDensify() layers and their idx which were not reset to -1. I tried to reset during the check: 
```
            dc_model.reset_layers() 
            for mod in dc_model.modules():
                try:
                    if mod.idx != -1: 
                        mod.reset()
                        print(mod.idx)
                except AttributeError:
                    continue 
```
but this does not seem to work, the idx is still not reset to -1. How should I reset these densify layers? 

2. I am sorry but I am having some difficulties understanding your explanation. In my case, I am trying to have one instance of point-wise convolution, followed by 5 unique instances of densify layers for the 5 outputs (which is currently giving me wrong results). 
"you'd need to apply the bias the first 5 times" -> does this mean bias=True for the 5 instances of pw convolution layers in the first frame? 
"When you use 5 instances of the layer, all of them are only called once in the first frame." -> are you referring to if I use 5 instances of the pw convolution layer, bias=True are only called once in the first frame?
"Can you please check if this layer uses a bias?" -> yes, the last pw convolution layer has bias=True. 
Which layers should I reset to frame_idx to -1 before the 5 calls (of densify layers?) in the first frame? 

Thank you!

**Comment by dabeschte** [comment]
1: DCDensify's reset function does not reset the idx ([See](https://github.com/facebookresearch/DeltaCNN/blob/07f49061483f888752df842b461d8a85ed7c062a/src/deltacnn/sparse_layers.py#L1332) , that is a mistake on my end. But it also does not matter, because these indices are afair only used for tensorboard logging. i think the resetting is done correctly here. You might be missing something else here. Did you already try out tensorboard logging? That should give you an indication an indication at which point the output differs. You could e.g. first infer sequence a and then b individually and the after ech other.

2: what I meant is that DeltaCNN only applies the bias in the first frame. With every "forward" call of a layer, I increment the "frame_idx" variable to count if I already applied the bias or not. Woth frame_idx >= 1, I won't apply the bias another time. So when you use one instance of a pw layer 5 times in the same frame, only the first output will be correct, but the other outputs will miss the bias.  When you use 5 instances of the pw layer, it works out of the box. If you only like to use a single instance, you must take care that the bias is applied on the other 4 outputs as well by manually resetting the "frame_idx" to -1 for those layers that you use multiple times (in your case, the pw conv layer). I hope it is clear now. If not, please let me know .
```
for x in x_list:
   if first_frame:
       pw_layer.frame_idx = -1
   out = pw_layer(x)
```

**Comment by kai981** [comment]
Hi, some updates :)

1: After some checks I think I found the problem. So it's not a reset model issue, but a problem with the batch norm layers in the Head layer, and the results are actually different even in the first seq run, but the differences are not as large to cause a major AP drop in the end so I didn't spot the problem in the first seq previously. The bn layers are constructed as such:
```
        self.bn_rep = nn.ModuleList()
        for _ in range(3): 
            self.bn_rep.append(nn.ModuleList([
                nn.Sequential(OrderedDict([('bn', DCBatchNorm2d(fpn_channels, **config.norm_kwargs))]))
                for _ in range(5)]))
```
and the forward runs as such:
```
        outputs = []
        for level, predict in enumerate(self.predict_rep):  
            x_level = x[level]
            for conv, bn, act in zip(self.conv_rep, self.bn_rep, self.act_rep):
                x_level = conv(x_level)
                x_level = bn[level](x_level)  
                x_level = act[level](x_level)
            x_level = predict(x_level)
            outputs.append(x_level)
        return outputs
```
The results after the conv layer is still correct (exact the same at least up to 4 dp), but starts to differ after the bn layer. I don't think it's an issue of DCBatchNorm2d itself since I am also using bn layers before this and the results until the conv layer is still fine. Maybe it's the way the bn layers are constructed? Do you have any idea? 

2. Thanks for your patience :) Yes, I get what you mean now, and will try the above correction!

Thank you!

**Comment by dabeschte** [comment]
Hey!

Happy to hear that you had some progress. And sorry for the whole project to be so confusing. Switching from PyTorch to DeltaCNN should be a bit simpler, but I did not take this case of layer reuse into account and this seems to cause a lot of issues here.
If I understand your code correctly, bn_rep is a list of shape bn_rep[3][5] which means that you apply 5 batch normalizations in the line `x_level = bn[level](x_level)` instead of only one. This might already be the issue. And since batch normalization is only a operation `y = x * scale + bias`, the bias is only applied onto the first output again. What I think you actually wanted to do is the following:

```
        outputs = []
        for level, predict in enumerate(self.predict_rep):  
            x_level = x[level]
            for step_idx, (conv, bn, act) in enumerate(zip(self.conv_rep, self.bn_rep, self.act_rep)):
                x_level = conv(x_level)
                x_level = bn[level][step_idx](x_level)  
                x_level = act[level][step_idx](x_level)
            x_level = predict(x_level)
            outputs.append(x_level)
        return outputs
```
This way, every b_norm layer is working independently and only 1 b_norm is applied per output. If you instead wanted to only instantiate only a single b_norm per level (b_rep of shape [3]), you'd again need the fix suggested for reusing the same convolutional layer with bias multiple times.

```
        outputs = []
        for level, predict in enumerate(self.predict_rep):  
            x_level = x[level]
            for step_idx, (conv, bn, act) in enumerate(zip(self.conv_rep, self.bn_rep, self.act_rep)):
                x_level = conv(x_level)
                if first_frame:
                     bn[level].reset()
                x_level = bn[level](x_level)  
                x_level = act[level][step_idx](x_level)
            x_level = predict(x_level)
            outputs.append(x_level)
        return outputs
```
In all cases, you need 15 instances of the activation layer (act_rep should be of shape [3][5]).
Also, if you are using a batch normalization after a convolution, you can leave the bias away from the convolutional layer - and therefore don't have to worry about resetting the convolutional layer.

I hope this works!

**Comment by kai981** [comment]
Hi, no worries I am really learning a lot along the way :)

Yes bn_rep is bn_rep[3][5], but since zip is used, so bn will just be bn[5] right? Or zip shouldn't be used here?
And yes the conv layers do not use biases. 

Thank you!

**Comment by dabeschte** [comment]
yes, bn will only be of shape [5]. But that means that you are applying 5 batch normalizations on every input. You'd only want to apply a single batch normalization in every iteration. But try my fix, it should work - > you'd have to use the same concept also during training. But this seems like a bug to me

**Comment by kai981** [comment]
Even if it's indexed with bn[level] it's still applying 5 batch normalizations? 
I am using the same way to construct bn_rep, and tried your suggestion to index as bn[level][step_idx], however, I encountered an error `IndexError: index 1 is out of range` for the line ` x_level = bn[level][step_idx](x_level)`. 
Or should I change the way bn_rep is constructed?

**Comment by dabeschte** [comment]
Ah sorry, I apologize. I confused bn with bn_rep here. As you said, bn will be of shape [5] and that is of course true.

Now I think it is actually correct - please forget my previous comment about that part of the code. 
But are you using the same batch normalization layers also in the "original" model that you use during training? If not, it is very likely that many of the bnorm layers are simply not correctly initialized, or not initialized at all.
bnorm layers contain parameters, so you can't simply use more bnorm layers in DeltaCNN than in the original model.

**Comment by kai981** [comment]
Hi, some updates and new questions :)

Your suggestion gave me some ideas to debug and finally managed to find the issue. It's actually because of some reset mechanism in the Head layer that changed the weights and biases of the bn layers. After fixing that, it's able to produce the correct results now. And I have some new questions. 

1. For the results in Table 2 of the supplementary material, are all results (cudnn, 'ours dense', 'ours sparse') obtained with fused batch norm and convolution layers? Do you have a rough estimate of how much does this fusing improve performance, in terms of FPS especially for RTX 3090? The current speedup I have is only about 1.3/1.4 as compared to the speedup of 1.9 in Table 2 for efficientdet_d1 MOT16. This is with the default threshold of 0 for all layers and no fused bn and conv layers. 

2. In terms of setup to obtain results for AP and FPS in Table 2, the only difference is just the way input image is processed? So for AP, it's resize and fill up rest with black pixels, for FPS, it's resize and crop. And the other things are the same? For instance, tuned thresholds and fused bn and conv layers, etc. 

3. It was stated in the paper that having high threshold for the first layer (the sparsify layer if i am not wrong) and a dilation of 7 pixels resulted in 8% speedup and slightly higher accuracy on MOT16. Is this result obtained from model with fused bn and conv layers and also tuned thresholds? I have tried to put threshold of 0.5 and dilation 7 pixels for the first sparsify layer, with no fused bn & conv layers and other thresholds are still default to 0, this does not seem to have any obvious speedup (or maybe it's rather minor so I will try to experiment a few more times) but resulted in slight drop of AP. 

4. Is there any reason why Tensor Cores are not supported? I am just curious so if it's too complicated to explain then it's alright you can skip this question :)

5.  This is also a side question. I do notice that different machines will cause FPS to vary even with the same setup and model, but not familiar with GPU or CUDA enough to know why. The FPS that I have obtained is just a single digit for both original model and dc model, whereas the FPS for RTX 3090 in Table 2 for efficientdet_d1 MOT16 is 23.3 and 45.0. Just wondering if there is any special setup that you happen to remember and could be shared? 

Thank you!

**Comment by dabeschte** [comment]
That's great to hear!

1: yes, I always used the same NN (with same fused layers), and only changed DCConv2d.backend and the thresholds respectively. I am not sure how much that would save exactly. But I figured during my evaluations that I need to optimize general performance to see larger gains when optimizing the convolution. When the network is limited by the cpu or when too many small layers are running, the speedup is lower. Do you reach approximately the same frame rates? I assume that PyTorch was mainly CPU limited in this case. And you probably have a faster CPU than I (I only had a intel 7700) and are therefore less cpu limited. I guess you see similar frame rates for DeltaCNN, but greater frame rates for cuDNN? that would confirm my theory.

2: correct

3: yes, I am always using the same model with fused bn/conv and tuned thresholds. I understand that you can't see s performance gain when the remaining thresholds are 0. But try using, e.g. 0.1 for all layers and then compare using a threshold of 0.5 on the first layer with dilation compared to only using e.g, a threshold of 0.1 without dilation. Try the same with 0.15 for all layers if the difference is not large enough. You should see a great speedup with roughly the same accuracy.

4: I honestly just did not have the time and both the Jetson Nano and the GTX 1050 do not support them. It should be quite straight forward to apply our tile skipping approach to tensor core accelerated convolutions. But I already had to implement all layers in 32bit and 16 bit mode which was way too much work for a paper :D

5: Do you only get single digit FPS with a 3090? what is your CPU? And do you run the code with debugging disabled (if you start it from an IDE)? Do you at least get comparable results when you increase the batch size? I unfortunately don't remember the gains from fusing layers, but for b=1, it is definitely significant. Other problems could be some synchronizations between gpu and cpu which are slow (e.g. print(loss) or x.cpu() or something like that). If you send me your email address, I can give you my code that I submitted with the paper. But I don't want to host it publicly on Github because of licenses and missing time to clean up the code.

**Comment by kai981** [comment]
Hi! 
Thank you so much for the suggestions, I guess there's more debugging to do. 
And yes I would really appreciate it if I could have a look at the code to help me figure things out! 
My email address is: luo_kaiwen98@hotmail.com
